### PR TITLE
drivers: stm32 qspi flash driver enables the MemoryMapped mode

### PR DESCRIPTION
**Pull request by FRASTM**
This PR is enabling the MemoryMapped to access the qua NOR flash at the end of the initialization.
With this configuration, the NOR flash access is done with memcopy in read or write
Note that the erase is not supported but writing is possible in MemoryMapped mode.
(It is not recommended to fill the memory with 0xFFFFFFFF)

"QSPI NOR devices are often used for read-only applications in memory mapped mode. And when the program instructions are executed directly from the flash memory, this is known as Execute in Place or XIP mode."

The drivers detects the  CONFIG_STM32_MEMMAP  flag to enable the MemoryMapped mode.

Requires PR [#61082](https://github.com/zephyrproject-rtos/zephyr/pull/62349)

### CODE REVIEW

**Comment by FRASTM** [comment]
After running the samples/drivers/spi_flash, reading the quad NOR flash of the stm32h747_disco_m7 in MemoryMapped mode gives the 
```
Perform test on single sector


Data read : 
000ff000  read 55
000ff001  read aa
000ff002  read 66
000ff003  read 99
000ff004  read ff
000ff005  read ff
000ff006  read ff
000ff007  read ff
```

**Comment by FRASTM** [comment]
reabse on e9dd938821e6cd0024cf5fcc50da1d3b7300b8f4

File: soc/arm/st_stm32/stm32h7/mpu_regions.c [hunk 1]
```diff
@@ -9,9 +9,11 @@
 
 static const struct arm_mpu_region mpu_regions[] = {
 	MPU_REGION_ENTRY("FLASH", CONFIG_FLASH_BASE_ADDRESS,
-					 REGION_FLASH_ATTR(REGION_FLASH_SIZE)),
+					REGION_FLASH_ATTR(REGION_FLASH_SIZE)),
 	MPU_REGION_ENTRY("SRAM", CONFIG_SRAM_BASE_ADDRESS,
-					 REGION_RAM_ATTR(REGION_SRAM_SIZE)),
+					REGION_RAM_ATTR(REGION_SRAM_SIZE)),
+	MPU_REGION_ENTRY("QSPI", 0x90000000,
+					REGION_EXTMEM_ATTR(REGION_256M)),
```
Comment (GeorgeCGV) [inline soc/arm/st_stm32/stm32h7/mpu_regions.c:16]: What is the point of having it in `stm32h7.dtsi` as `RAM` and here as `EXTRAM`?

File: dts/arm/st/h7/stm32h7.dtsi [hunk 2]
```diff
@@ -46,9 +46,10 @@
 
 	quadspi_memory: memory@90000000 {
 		compatible = "zephyr,memory-region", "mmio-sram";
+		device_type = "memory";
 		reg = <0x90000000 DT_SIZE_M(256)>;
 		zephyr,memory-region = "QSPI";
-		zephyr,memory-attr = "EXTMEM";
+		zephyr,memory-attr = "RAM";
```
Comment (GeorgeCGV) [inline dts/arm/st/h7/stm32h7.dtsi:52]: `REGION_RAM_ATTR` (`RAM`) - prohibits execution, doesn't it?

**Comment by FRASTM** [comment]
New version where the HAL is not changed thanks to direct access to the hqspi instance in the flash_stm32_qspi driver
The flash base address and size are directly placed in the flash-controller node of the Device Tree : reg = <0x90000000 DT_SIZE_M(64)>; (no more size property required)

Tested with the stm32h747i_disco_m7 disco board ; using a sample to demonstrate the read operation in memorymapped mode (no erase)

File: drivers/flash/flash_stm32_qspi.c [hunk 3]
```diff
@@ -360,6 +361,49 @@ static bool qspi_address_is_valid(const struct device *dev, off_t addr,
 	return (addr >= 0) && ((uint64_t)addr + (uint64_t)size <= flash_size);
 }
 
+#ifdef CONFIG_STM32_XIP
+
+/* Function to return true if the quad-NOR flash is in MemoryMapped else false */
+static bool qspi_is_memorymapped(const struct device *dev)
```
Comment (erwango) [inline drivers/flash/flash_stm32_qspi.c:367]: According to HAL team, we should be able to use the HAL_GetStatus API.

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/arm/st/h7/stm32h7.dtsi (hunk 2 above)
Comment (GeorgeCGV) [reply dts/arm/st/h7/stm32h7.dtsi:52]: Doesn't apply since #61577 merge.

**Comment by FRASTM** [comment]
Cannot read the QSPI flash correctly
(gdb) x /8x 0x90000000
0x90000000:	0x0000f33d	0x00360000	0x00000000	0x00000000
0x90000010:	0x00000000	0x00000000	0x00000000	0x00000000
0x90000020:	0x00000000	0x00000000	0x00000000	0x00000000
0x90000030:	0x00000000	0x00000000	0x00000000	0x00000000

when the STM32CubeProgrammer gives 
0x90000000:	96F3B83D	00000000	00000400	00003B36

File: drivers/flash/flash_stm32_qspi.c (hunk 3 above)
Comment (erwango) [reply drivers/flash/flash_stm32_qspi.c:367]: > According to HAL team, we should be able to use the HAL_GetStatus API.

Update: This won't work as `HAL_GetStatus()` is based on driver internal variable (reset when switching from MCUBoot to Zephyr application), not hw setting

**Comment by FRASTM** [comment]
tested with stm32l496g_disco : XiP is possible for the external NOR quad-flash
```
*** Booting Zephyr OS build zephyr-v3.4.0-3899-g31d51386c77d ***
I: Starting Direct-XIP bootloader
I: Image 0 Primary slot: Image not found
I: Secondary slot: version=0.0.0+0
I: Image 0 loaded from the secondary slot
I: Bootloader chainload address offset: 0x0
I: Jumping to the image slot
*** Booting Zephyr OS build zephyr-v3.4.0-3899-g31d51386c77d ***
Hello World! from external flash  stm32l496g_disco      
```

**Comment by FRASTM** [comment]
tested with  stm32h747i_disco_m7, 
```
*** Booting Zephyr OS build zephyr-v3.4.0-3899-g31d51386c77d ***
I: Starting Direct-XIP bootloader                       
I: Image 0 Primary slot: Image not found                
I: Image 0 Secondary slot: Image not found
I: No slot to load for image 0
E: Unable to find bootable image
```

**Comment by FRASTM** [comment]
tested with stm32f469i_disco : XiP is possible for the external NOR quad-flash  
Note that dummy cycle is 10 for this board compared to 6 for stm32l496g_disco

```
*** Booting Zephyr OS build v3.5.0-rc2-65-ged053ae2f3d8 ***
I: Starting Direct-XIP bootloader
I: Image 0 Primary slot: Image not found
I: Secondary slot: version=0.0.0+0
I: Image 0 loaded from the secondary slot                               
I: Bootloader chainload address offset: 0x0                             
I: Jumping to the image slot                                            
*** Booting Zephyr OS build v3.5.0-rc2-64-g217d0764689e ***             
Hello World! from external flash  stm32f469i_disco      
```

**Comment by FRASTM** [comment]
The PR #68607 is now replacing this PR

**Pull request closed without merging**